### ISSUE TITLE
2-arity Goal argument for iterate/3?

### ISSUE THREAD

**Issue by mndrix** [mndrix/list_util#16]
During discussion with @eazar001 on #15, I realized that our current API for `iterate/3` is perhaps more complicated than it should be.  The `Goal` argument has arity 3.  That allows one to make a distinction between state values and returned values.  However, in most use cases the state value and the returned value are identical.

If `Goal` had arity 2, we'd get simpler code.  For example:

```prolog
integers(Z) :- iterate(succ,1,Z). % Z = [1,2,3,...]
repeat(X,Xs) :- iterate(=,X,Xs). % Xs = [X,X,X,...]
```

This seems more natural than our current API which requires an auxiliary predicate in each case.

For what it's worth, using `Goal` with arity 2 more closely emulates the successful [iterate :: (a -> a) -> a -> [a]](http://hackage.haskell.org/package/base-4.8.1.0/docs/Prelude.html#v:iterate) function from Haskell's Prelude.

**Comment by eazar001** [comment]
I agree that having `:Goal ` carry arity 2 makes things visually a lot easier.

**Comment by mndrix** [comment]
I'm having second thoughts about this.  We now have three examples of a 3-arity `Goal` being very useful:

  * [infinite Fibonacci](https://github.com/mndrix/list_util/blob/e48267353e70d7b62710cd4328b1eea71e81d1a9/t/iterate.pl#L24-L28)
  * [original repeat/2](https://github.com/mndrix/list_util/blob/f8fefda891f74818a08fd3109dee38f6b6a3c191/prolog/list_util.pl#L312-L313)
  * [cycle/2](https://github.com/mndrix/list_util/blob/master/prolog/list_util.pl#L324-L325)

In all three cases, the state that's being iterated is different than the values being produced.  I'd hate to lose this powerful ability unless we have to.

Maybe we could use reflection to get the 2-arity behavior when it's wanted.  Something like:

```prolog
iterate(Module:Goal,State,List) :-
    functor(Goal,Name,Arity),
    ( N is Arity+3, current_predicate(Module:Name/N) ->
        iterate_(Goal,State,List)
    ; N is Arity+2, current_predicate(Module:Name/N) ->
        iterate_(iterate_wrapper(Goal),State,List)
    ; otherwise ->
        throw(iterate_goal_bad_arity)
    ).

iterate_wrapper(Goal,X0,X,X) :-
    call(Goal,X0,X).
```

**Comment by eazar001** [comment]
Incidentally, as I was working a little bit more with your `iterate/3`, I also started noticing just how powerful it was as well. Although your proposition seems like a good compromise, nonetheless.
